Add metric filter option to metrics transform processor

james-bebbington opened the pull request:
**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1088#issuecomment-706916609

**Description:**

Add a simple include filter option with `strict` or `regexp` match type so that transforms can be applied to a range of metrics. When using the `regexp` match type, `new_name` supports expansion.

This implementation is compatible with the config proposed [here](https://github.com/open-telemetry/opentelemetry-collector/issues/1081#issuecomment-707450870) so could be switched out for that implementation in the future if desired. Also retained the `metric_name` config option for backwards compatibility but removed references to this from the documentation.

Continues optimizing for the `strict` match case, i.e. retain `O(n) + O(t)` for `strict` matches where metric names are unique, at the cost of making `regexp` matching slightly slower. `regexp` matches will be `O(n) * O(t)`.

Also resolves an edge case where if multiple metrics have the same name only one of them would previously have been transformed.

**Documentation:**

Updated the documentation to specify how to use the `match_type` filter, and generally cleaned up the README to be more easily digested.

james-bebbington committed:
Add metric filter option to metrics transform processor

jpkrohling commented on processor/metricstransformprocessor/metrics_transform_processor.go:
```diff
@@ -27,10 +29,42 @@ import (
+}
+
+func (f internalFilterRegexp) getMatches(toMatch metricNameMapping) []*metricspb.Metric {
+	const capExpectedMatches = 10
```
Any reasoning for this value?

james-bebbington replied:
Just seemed like a sensible default. If a user is using `regex` match type, they probably expect more than 1 match.

jpkrohling replied:
Sounds good. Perhaps add this as a comment to the code? Otherwise, it's still going to be a magic number :)

jpkrohling commented on processor/metricstransformprocessor/metrics_transform_processor.go:
```diff
@@ -27,10 +29,42 @@ import (
+}
+
+func (f internalFilterStrict) getMatches(toMatch metricNameMapping) []*metricspb.Metric {
+	if metrics, ok := toMatch[f.include]; ok {
```
You can replace this whole func with: `return toMatch[f.include]`.

james-bebbington replied:
This comment has become redundant as I refactored this function to support substition - see comment below

jpkrohling commented on processor/metricstransformprocessor/metrics_transform_processor.go:
```diff
@@ -60,29 +119,26 @@ func (mtp *metricsTransformProcessor) ProcessMetrics(_ context.Context, md pdata
+					if transform.Action == Update {
+						nameToMetricMapping.remove(metricName, metric)
+					}
+					nameToMetricMapping.add(metric.MetricDescriptor.Name, metric)
```
Shouldn't this be `transform.NewName`?

james-bebbington replied:
The metric name has been transformed at this point so either would work.

I wanted to use `metric.MetricDescriptor.Name` here so this works even if `transform.NewName` supported substitution, which I've now added to this PR - see comment below.

james-bebbington committed:
Address review comments and add new_name regex substitution

james-bebbington commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,134 @@
+| Rename label values | For metric `system.cpu.usage`, label `state`, rename label value `idle` to `-` |
+| Delete label values | For metric `system.cpu.usage`, delete all data points where label `state` has value `idle` |
+| Toggle the data type of scalar metrics between `int` and `double` | For metric `system.cpu.usage`, change from `int` data points to `double` data points |
+| Aggregate across label sets by sum, mean, min, or max | For metric `system.cpu.usage`, retain the label `state`, but aggregate away the label `cpu` |
```
Reworded a few of these to make it clearer

james-bebbington commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,134 @@
-      aggregated_values: [values...]
-      new_value: <new_value> 
-      aggregation_type: {sum, mean, max}
+        # update_label action can be used to update the name of a label or the values of this label
```
Fixed

james-bebbington commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,134 @@
+        # label_set contains a list of labels that will remain after aggregation; if action is aggregate_labels, label_set is required
+        label_set: [labels...]
+        # aggregation_type defines how excluded labels will be aggregated
+        aggregation_type: {sum, mean, max}
```
`mean` = average

This list was actually missing `min` though which is probably what you meant. Added now 😄

james-bebbington commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,134 @@
+operations:
+  - action: add_label
+    new_label: version
+    new_value: opentelemetry collector {{version}}
```
Yes, as specified [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1447/files#diff-aa89510ae2d8f864249e5e975fce5d26d2f902913f8b14e5fe83c79ff9422b47R29).

james-bebbington commented on processor/metricstransformprocessor/README.md:
```diff
@@ -99,19 +139,37 @@ operations:
-# aggregate away everything but `state` using summation
+# delete the label value 'value' of the label 'label'
+...
+operation:
```
Added to all the examples

james-bebbington commented on processor/metricstransformprocessor/config.go:
```diff
@@ -14,9 +14,17 @@
 
 const (
+	// IncludeFieldName is the mapstructure field name for Include field
+	IncludeFieldName = "include"
```
Yea there are quite a few things that could be cleaned up with how the config is structured. I will leave this for a follow up PR though.

james-bebbington commented on processor/metricstransformprocessor/config.go:
```diff
@@ -36,6 +44,16 @@ const (
 
+const (
+	// StrictMatchType is the FilterType for filtering by exact string matches.
+	StrictMatchType = "strict"
```
As above

james-bebbington commented on processor/metricstransformprocessor/config.go:
```diff
@@ -46,8 +64,12 @@ type Config struct {
+	MetricIncludeFilter FilterConfig `mapstructure:",squash"`
+
+	// MetricName is used to select the metric to operate on.
+	// DEPRECATED
```
Updated comment to specify that `MetricIncludeFilter` should be used instead.

james-bebbington commented on processor/metricstransformprocessor/config_test.go:
```diff
@@ -71,7 +71,9 @@ var (
 				},
 				Transforms: []Transform{
 					{
-						MetricName: "old_name",
+						MetricIncludeFilter: FilterConfig{
```
Added `config_deprecated.yaml` test

james-bebbington commented on processor/metricstransformprocessor/factory.go:
```diff
@@ -128,7 +166,20 @@ func buildHelperConfig(config *Config, version string) []internalTransform {
+		}
+		return internalFilterRegexp{include: r}, nil
+	default:
+		return internalFilterStrict{include: filterConfig.Include}, nil
```
Validation actually occurs before this function runs so that shouldn't be possible. Nevertheless, I removed the `default` case so invalid values will result in an error instead of silently using `Strict`.

james-bebbington commented:
@jpkrohling FYI while addressing your comments, I also added one minor feature to this PR, which is that if the `match_type` is `regexp`, then `new_name` now supports regexp expansion, see the [updated test case](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1447/files#diff-55c1fe63fa33e79a5e074c8f9333a8cdfa9a31df913ac3a6052646ab7cfa4fc4R85).

jpkrohling commented on processor/metricstransformprocessor/metrics_transform_processor_testcases_test.go:
```diff
@@ -74,13 +76,44 @@ var (
+				{
+					MetricIncludeFilter: internalFilterRegexp{include: regexp.MustCompile("^(metric)(?P<namedsubmatch>[12])$")},
+					Action:              Update,
+					NewName:             "new/$1/$namedsubmatch",
```
Looks really cool! Tell people about this in the readme, looks like it's missing from there.

jpkrohling commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,155 @@
+| Aggregate across label values by sum, mean, min, or max           | For label `state`, sum points where the value is `user` or `system` into `used = user + system` |
+
+In addition to the above, when adding or updating a label value, specify
+`{{version}}` to include the application version number
```
Perhaps replace "application" (which can be ambiguous) with "this collector's"? 
```suggestion
`{{version}}` to include this collector's version number
```

jpkrohling commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,155 @@
-# rename the label cpu to core
+# for all system.cpu metrics, rename the label state to cpu_state
+include: ^system\.cpu\.
+action: update
```
Isn't this missing a ` match_type: regexp`?

jpkrohling commented on processor/metricstransformprocessor/README.md:
```diff
@@ -1,94 +1,155 @@
+operations:
+  - action: add_label
+    new_label: version
+    new_value: opentelemetry collector {{version}}
```
I might be a bit paranoid, but this case isn't part of the tests. Would you mind adding it?

james-bebbington replied:
The application version is known at initialization time so this replacement is actually done then, when mapping the config to internal transforms. There is a test for this [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/processor/metricstransformprocessor/factory_test.go#L158).

jpkrohling replied:
LGTM then
